src/features/weak-references.md

```diff
@@ -54,104 +57,138 @@ A JavaScript `WeakMap` is not really _weak_: it actually refers _strongly_ to it
-
-:::note
-**Note:** The finalization callback does not run immediately after garbage-collecting the image object. It either runs at some point in the future, or not at all — the spec doesn’t guarantee that it runs! Keep this in mind when writing code.
-:::
+The downside to this approach is that it is manual memory management. `MovingAvgComponent`, and all other users of the `MovingAvg` class, must remember to call `dispose` or suffer memory leaks. The application behavior doesn't depend on the event listener of this diagnostic class, and the listener is expensive in terms of memory use but not in computation. What we really want is for the listener's lifetime to be logically tied to the `MovingAvg` instance, so that `MovingAvg` could be used like any other JavaScript object whose memory is automatically reclaimed by the garbage collector.
 
-Here, we register a callback to remove keys from the cache when the image objects are garbage-collected:
+`WeakRef`s make it possible to solve the dilemma by creating a _weak reference_ to the actual event listener, and then wrapping that `WeakRef` in an outer event listener. This way, the garbage collector can clean up the actual event listener and the memory that it holds alive, like the `MovingAvg` instance and its `events` array.
 
 ```js
-const cache = new Map();
+function addWeakListener(socket, listener) {
+  const weakRef = new WeakRef(listener);
+  const wrapper = (ev) => { weakRef.deref()?.(ev); };
+  socket.addEventListener('message', wrapper);
+}
 
-const finalizationRegistry = new FinalizationRegistry((name) => {
-  const ref = cache.get(name);
-  if (ref !== undefined && ref.deref() === undefined) {
-    cache.delete(name);
+class MovingAvg {
+  constructor(socket) {
+    this.events = [];
+    this.listener = (ev) => { this.events.push(ev); };
+    addWeakListener(socket, this.listener);
   }
-});
+}
 ```
 
 :::note
-**Note:** The `ref !== undefined && ref.deref() === undefined` is required because we could’ve added a new `WeakRef` with the same `name` between the old `WeakRef` enqueueing the finalization callback and actually running the finalization callback.
+**Note:** `WeakRef`s to functions must be treated with caution. JavaScript functions are [closures](https://en.wikipedia.org/wiki/Closure_(computer_programming)) and strongly reference the outer environments which contain the values of free variables referenced inside the functions. These outer environments may contain variables that _other_ closures reference as well. That is, when dealing with closures, their memory is often strongly referenced by other closures in subtle ways. Keep this in mind when writing code.
```

Comment:
Say explicitly that this is why we have a separate `addWeakListener` function, and that it's explicitly not local to the MovingAvg constructor to not be part of its lexical scope.

Reply:
Will add.